distributed/tensor_parallelism/two_d_parallel_example.py

```diff
@@ -54,74 +53,109 @@
 """
 
 
-def demo_2d(rank, args):
+def demo_2d(args):
     """
     Main body of the demo of a basic version of tensor parallel by using
     PyTorch native APIs.
     """
-    print(f"Running basic Megatron style TP example on rank {rank}.")
-    setup(rank, args.world_size)
+    torchrun_setup()
+
+    # understand world topology
+    _rank = int(os.environ["RANK"])
+    _local_rank = int(os.environ["LOCAL_RANK"])
+    _world_size = int(os.environ["WORLD_SIZE"])
+    _local_world_size = int(os.environ["LOCAL_WORLD_SIZE"])
+
+    torch.cuda.set_device(_local_rank)
+
+    def rank_print(msg):
+        """helper function to print only on global rank 0"""
+        if _rank==0:
+            print(f"{msg}")
+
+    print(f"Running basic Megatron style TP example on rank {_rank}.")
+
     assert (
-        args.world_size % args.tp_size == 0
-    ), "World size needs to be divisible by TP size"
+        _world_size % args.tp_size == 0
+    ), f"World size {_world_size} needs to be divisible by TP size {args.tp_size}"
+
+    device = f"cuda"
 
     # create a sharding plan based on the given world_size.
-    device_mesh = DeviceMesh(
-        "cuda", torch.arange(0, args.world_size).view(-1, args.tp_size)
-    )
 
-    # create model and move it to GPU with id rank
-    model = ToyModel().cuda(rank)
-    # Create a optimizer for the parallelized module.
-    LR = 0.25
-    optimizer = torch.optim.SGD(model.parameters(), lr=LR)
-    # Parallelize the module based on the given Parallel Style.
-    parallel_style = SequenceParallel() if args.run_seq_parallel else PairwiseParallel()
-    model = parallelize_module(model, device_mesh, parallel_style, tp_mesh_dim=1)
-
-    # We need to register hooks for TP + FSDP integration.
-    assert (
-        enable_2d_with_fsdp()
-    ), "FSDP 2D hook is not registered. Please use PyTorch with version >= 2.0"
+    dp_size = _world_size // args.tp_size
+
+    # Create a device mesh with 2 dimensions.
+    # First dim is the data parallel dimension
+    # Second dim is the tensor parallel dimension.
+    device_mesh = init_device_mesh(device, (dp_size, args.tp_size), mesh_dim_names=("dp","tp"))
+    assert device_mesh is not None, "unable to create valid device mesh"
```

Comment:
Not against having this assertion, but just curious of why we need this assertion. Have you run into scenario where device_mesh is not successfully created but it silently passed?

Reply:
this was just purely defensive programming, reflex add :) 
Per Wanchao init_device_mesh will internally assert assuming failure to make a device mesh, so have removed as it's thus a redundant check that would never be hit.